docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,429 @@
+
+### Extensions
+
+#### OS extensions
+
+The OS extensions are responsible for populating the `inPlaceUpdateConfig` field within the `OperatingSystemConfig` status. This configuration specifies the command to be executed by the `gardener-node-agent` to perform the update. The command could be invoking an inbuilt utility/tool or a custom script.
+
+### Worker pool hash calculations
+
+The worker pool hash calculation is derived from the rolling update triggers detailed [here](https://github.com/gardener/gardener/blob/master/docs/usage/shoot-operations/shoot_updates.md#rolling-update-triggers).
+
+For in-place updates, the following fields are excluded from the hash calculation to ensure that the secret name for the `gardener-node-agent` remains unchanged:
+
+- `.spec.kubernetes.version` or `.spec.provider.workers[].kubernetes.version`
+- `.spec.provider.workers[].machine.image.version`
+- `.status.credentials.rotation.{certificateAuthorities,serviceAccountKey}.lastInitiationTime`
+- `.spec.kubernetes.kubelet.{kubeReserved,systemReserved,evictionHard,cpuManagerPolicy}` or `.spec.provider.workers[].kubernetes.kubelet.{kubeReserved,systemReserved,evictionHard,cpuManagerPolicy}`
+
+However, these fields are included in the [`WorkerPoolHashV2` function](https://github.com/gardener/gardener/blob/81e6fd096f8725d0b374315fe134cb35f604bd68/extensions/pkg/controller/worker/machines.go#L163-L175) to ensure changes to the `MachineClass` are properly reflected.
+
+### Failures and Recovery Strategy
+
+In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption. The error will be reported in the `node.machine.sapcloud.io/update-failure-message` annotation on the node which can be used subsequently by the Shoot care controller.
+
+#### Types of Failures During Update
+
+##### Drain Timeout/Failure
+
+- A node is marked as `node.machine.sapcloud.io/ready-for-update` only after a successful drain.
+- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
+- If a drain operation times out or fails, the node will be forcefully drained.
+
+##### Kubernetes Update Failures
+
+- **Update Halt on Failure:** `gardener-node-agent` will label the failed node with an `node.machine.sapcloud.io/update-failed` label. This label signals `machine-controller-manager` to stop further updates when the number of unhealthy machines reaches configured `maxUnavailable`.
```

Comment:
What does `Kubernetes Update Failures` mean or in other words what does `gardener-node-agent` check to signal this failure?
I can image "easy" failures, e.g. `kubelet` failing to start, and "harder-to-detect" failures, e.g. pods not being started correctly and hence failing.
Could you please share what you intend to do here?

Reply:
After the update if GNA shows the kubelet status as unhealthy for a certain period then we will consider that the kubelet update failed.

> "harder-to-detect" failures, e.g. pods not being started correctly and hence failing.

For this, I think it is not possible.  A similar problem is present with `AutoReplaceUpdate` which we use as of today. If on the new node pods don't come up properly we can't do much.